File: csharp/src/Codec.cs
using System.Collections.Generic;

namespace codech
{
    internal record Codec(Matrix matrix)
    {
        public IEnumerable<byte> Encode(IEnumerable<byte> stream)
        {
            foreach (byte b in stream)
            {
                var (b1, b2) = this.matrix.Encode(b);
                yield return b1;
                yield return b2;
            }
        }

        public IEnumerable<byte> Decode(IEnumerable<byte> stream)
        {
            var enumerator = stream.GetEnumerator();
            while (enumerator.MoveNext())
            {
                var b1 = enumerator.Current;
                enumerator.MoveNext();
                var b2 = enumerator.Current;
                yield return this.matrix.Decode(b1, b2);
            }
        }
    }
}

File: java/src/io/jrimbault/codech/Codec.java
package io.jrimbault.codech;

public class Codec {
    private static final byte MASK = 0x0F;
    private final Matrix matrix;

    public Codec(Matrix matrix) {
        this.matrix = matrix;
    }

    public byte[] encode(byte[] stream) {
        return encode(matrix, stream);
    }

    public byte[] decode(byte[] stream) {
        return decode(matrix, stream);
    }

    private static byte[] encode(Matrix matrix, byte[] stream) {
        var list = new byte[stream.length * 2];
        for (var i = 0; i < stream.length; i += 1) {
            list[i * 2] = matrix.encode[stream[i] & MASK];
            list[i * 2 + 1] = matrix.encode[(stream[i] >> 4) & MASK];
        }
        return list;
    }

    private static byte[] decode(Matrix matrix, byte[] stream) {
        var list = new byte[stream.length / 2];
        for (var i = 0; i < stream.length; i += 2) {
            var a = stream[i];
            var b = stream[i + 1];
            var p1 = matrix.decode[a & 0xFF];
            var p2 = matrix.decode[b & 0xFF] << 4;
            list[i / 2] = ((byte) (p1 | p2));
        }
        return list;
    }
}
